From 968bc37469514daac816aa0fc52a28b2c4aa9e3f Mon Sep 17 00:00:00 2001
From: Lei Xu <lei@apache.org>
Date: Wed, 7 Oct 2015 14:33:43 -0700
Subject: [PATCH 1085/2748] HDFS-9176. Fix TestDirectoryScanner#testThrottling
 often fails. (Daniel Templeton via lei)

(cherry picked from commit 6dd47d754cb11297c8710a5c318c034abea7a836)
(cherry picked from commit ad1f0f307b4df5c868c34e297648ff10ee5885d8)

Change-Id: Iaec74124653dee34c86170b3a7c97fb2508af4c4
---
 .../hdfs/server/datanode/TestDirectoryScanner.java |   17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestDirectoryScanner.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestDirectoryScanner.java
index c170010..58d0985 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestDirectoryScanner.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestDirectoryScanner.java
@@ -652,17 +652,24 @@ public void testThrottling() throws Exception {
           interruptor.schedule(new Runnable() {
             @Override
             public void run() {
-              scanner.shutdown();
               nowMs.set(Time.monotonicNow());
+              scanner.shutdown();
             }
           }, 2L, TimeUnit.SECONDS);
 
           scanner.reconcile();
           assertFalse(scanner.getRunStatus());
-          LOG.info("Scanner took " + (Time.monotonicNow() - nowMs.get())
-              + "ms to shutdown");
-          assertTrue("Scanner took too long to shutdown",
-              Time.monotonicNow() - nowMs.get() < 1000L);
+
+          long finalMs = nowMs.get();
+
+          // If the scan didn't complete before the shutdown was run, check
+          // that the shutdown was timely
+          if (finalMs > 0) {
+            LOG.info("Scanner took " + (Time.monotonicNow() - finalMs)
+                + "ms to shutdown");
+            assertTrue("Scanner took too long to shutdown",
+                Time.monotonicNow() - finalMs < 1000L);
+          }
 
           ratio =
               (float)scanner.timeWaitingMs.get() / scanner.timeRunningMs.get();
-- 
1.7.9.5

